### PR TITLE
Refactor LS Workspace Manger 

### LINKED ISSUE #39097: [Improvement]: Notify user when trying to create a new ballerina source file with an existing name
### Description

The existing project structure doesn't allow having documents with the same name in the same module due to being loaded from a single file path. e.g. mod1 can only have one mod1.bal.

But, with the generated sources support added with https://github.com/ballerina-platform/ballerina-lang/pull/38977, the file structure doesn't restrict this. e.g. mod1 can have `modules/mod1/mod1.bal` and `generated/mod1/mod1.bal`.

### Describe your problem(s)

This is problematic because the document path is similar for such files and Ballerina internally relies on the document path to uniquely identify documents, for diagnostics, etc.
While source generation tools should create files with unique names, we need to notify users if they try to create source files with an existing generated source name. 

### Describe your solution(s)

From, project API perspective, if 2 files with the same name exist, a ProjectAPIException will be thrown and it will get logged in the output console for VSCode. But, we need to be sure that user notices what went wrong and notify on how to resolve preferably via a popup dialog.

### Related area

-> Other Area

### Related issue(s) (optional)

https://github.com/ballerina-platform/ballerina-lang/issues/38766

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

@IMS94 @malinthar

### PR DESCRIPTION
**Pull request by malinthar**
## Purpose
$subject

Fixes #39097 

## Approach

With the introduction of the `generated` module there is a risk of having duplicate files (same file path internally) in a project as depicted in the following image. 

image
 
**When a user creates a duplicate document, the user should be notified about it.** 

The ProjectAPI is able to identify the duplicate document when a project is created (loaded) using the [ProjectLoader.loadProject() method](https://github.com/ballerina-platform/ballerina-lang/blob/4af6439817b3a22dcee8365c1f40b4eef6ab4aba/compiler/ballerina-lang/src/main/java/io/ballerina/projects/directory/ProjectLoader.java#L44).

 If there are duplicate files the ProjectAPI throws an ProjectException.

imageThe ProjectAPI throws an ProjectException

This PR refactors the workspace manager to create a new project if the project is erroneous or a new file is created so that the ProjectAPI is able to identify and notify the user about duplicate files. 

## Samples
projectAPI

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java [hunk 1]
```diff
@@ -1242,32 +1254,43 @@ private void updateBalToolToml(String content, ProjectContext projectContext, bo
...
-                    // Reload the project
-                    projectContext.setProject(
-                            createProject(filePath, LSContextOperation.TXT_DID_OPEN.getName()).project());
-                    document = document(filePath, projectContext.project(), null);
-                } else {
-                    throw new WorkspaceDocumentException("Document does not exist in path: " + filePath.toString());
-                }
+                throw new WorkspaceDocumentException("Document does not exist in path: " + filePath.toString());
             }
+            Document updatedDoc = document.get().modify().withContent(content).apply();
```
Comment (KavinduZoysa) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java:1267]: This is not used

**Comment by KavinduZoysa** [comment]
Can you check the renaming scenario? i.e. rename a file in the generated folder into an existing file in the root folder.

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/workspace/TestWorkspaceManager.java [hunk 2]
```diff
@@ -553,12 +601,12 @@ public void testWorkspaceProjects() throws WorkspaceDocumentException,
         Path singleFileProject = workspacePath.resolve("workspace2").resolve("single.bal");
         Path project2File = workspacePath.resolve("workspace3").resolve("project2").resolve("main.bal");
         Path project3File = workspacePath.resolve("workspace3").resolve("project3").resolve("main.bal");
-      
-        
+
+
```
Comment (mindula) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/workspace/TestWorkspaceManager.java:605]: Let's remove this extra new line

File: language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/workspace/TestWorkspaceManager.java [hunk 3]
```diff
@@ -610,6 +658,15 @@ public Object[] fileOpenUpdateTestDataProvider() {
         };
     }
 
+    @DataProvider
+    public Object[] fileOpenWithDuplicateFilesDataProvider() {
+        return new Path[]{
+                RESOURCE_DIRECTORY.resolve("pkg_with_duplicate_files1").resolve("main.bal").toAbsolutePath(),
+                RESOURCE_DIRECTORY.resolve("pkg_with_duplicate_files2").resolve("main.bal").toAbsolutePath()
+
```
Comment (mindula) [inline language-server/modules/langserver-core/src/test/java/org/ballerinalang/langserver/workspace/TestWorkspaceManager.java:666]: Let's remove this new line

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java (hunk 1 above)
Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/workspace/BallerinaWorkspaceManager.java]: Fixed

File: language-server/modules/langserver-core/src/test/resources/project/pkg_with_duplicate_files2/generated/mod1/mod1.bal [hunk 4]
```diff
@@ -0,0 +1 @@
+string s = 20;
```
Comment (mindula) [inline language-server/modules/langserver-core/src/test/resources/project/pkg_with_duplicate_files2/generated/mod1/mod1.bal:1]: Add missing new lines

**Review by mindula — APPROVED** [review]

**Review by KavinduZoysa — APPROVED** [review]

**Pull request merged**